Use amend instead of commit for intermediate writes on temp branches (#584)
dynamical-org/reformatters: docs/parallel_processing.md

All metadata and chunk writes happen on a temporary branch (`_job_{job_name}`). Readers on `main` are unaffected. The flow:

1. **Worker 0 setup** — creates a temp branch from main's current snapshot, copies expanded metadata from the local tmp store, **commits** on the branch. This must be a commit (not amend) so the original main snapshot is preserved as the parent of the new work.
2. **All workers** — open sessions on the temp branch, write chunk data, **amend** the temp branch tip (via `amend_if_icechunk`, which falls back to `rebase` + `amend` on `ConflictError`). Each amend replaces the previous tip, so per-worker writes collapse into a single snapshot rather than each becoming its own snapshot.

> 3. **Last worker finalization** — if `update_template_with_results` produced a template that differs from the input (e.g. some reads failed and the time dim was trimmed), writes the new metadata on the branch and **amends** it into the existing tip (via `amend_if_icechunk`). If the template is unchanged, the temp tip already has the right metadata, so the rewrite + amend is skipped. Either way, finalize then atomically resets `main` to the temp branch tip using `reset_branch("main", snapshot, from_snapshot_id=original)`. This branch reset is what makes all writes visible to readers. The `from_snapshot_id` check ensures no concurrent process moved main.

Author: ```suggestion
3. **Last worker finalization** — if `update_template_with_results` produced a template that differs from the input (e.g. the time dim was trimmed or a coordinate or attribute was updated), writes the new metadata on the branch and **amends** it into the existing tip (via `amend_if_icechunk`). If the template is unchanged, the temp tip already has the right metadata, so the rewrite + amend is skipped. Either way, finalize then atomically resets `main` to the temp branch tip using `reset_branch("main", snapshot, from_snapshot_id=original)`. This branch reset is what makes all writes visible to readers. The `from_snapshot_id` check ensures no concurrent process moved main.
```
